webseed/request.go

```diff
@@ -12,28 +12,56 @@ import (
 
 // Escapes path name components suitable for appending to a webseed URL. This works for converting
 // S3 object keys to URLs too.
+// Contrary to the name, this actually does a QueryEscape, rather than a
+// PathEscape. This works better with most S3 providers. You can use
+// EscapePathWithCustomEncoding for a custom encoding.
 func EscapePath(pathComps []string) string {
+	return escapePath(pathComps, nil)
+}
+
+func escapePath(pathComps []string, encodeUrl func(string) string) string {
+	if encodeUrl == nil {
+		encodeUrl = url.QueryEscape
+	}
 	return path.Join(
 		func() (ret []string) {
 			for _, comp := range pathComps {
-				ret = append(ret, url.QueryEscape(comp))
+				ret = append(ret, encodeUrl(comp))
 			}
 			return
 		}()...,
 	)
 }
 
-func trailingPath(infoName string, fileComps []string) string {
-	return EscapePath(append([]string{infoName}, fileComps...))
+func trailingPath(infoName string, fileComps []string, encodeUrl func(string) string) string {
+	return escapePath(append([]string{infoName}, fileComps...), encodeUrl)
 }
 
 // Creates a request per BEP 19.
 func NewRequest(url_ string, fileIndex int, info *metainfo.Info, offset, length int64) (*http.Request, error) {
+	return newRequest(url_, fileIndex, info, offset, length, nil)
+}
+
+func NewRequestWithCustomUrlEncoding(
```

Comment:
What about `NewRequestWithOpts` and a `NewRequestOpt` struct that contains `EscapePath PathEscaper`?

Reply:
Done https://github.com/afjoseph/torrent/commit/a79c3bd5d30d99fb33ab7c5334c7d017884d4403

Reply:
I know I mentioned that NewRequest was public API, but at the cost of enumerating all the option styles for creating a Request, and not being able to find anywhere that it's actually used publicly, I've made it private instead.